llama-cpp-2/src/token/data_array.rs

```diff
@@ -47,4 +57,320 @@ impl LlamaTokenDataArray {
     {
         Self::new(data.into_iter().collect(), sorted)
     }
+
+    #[must_use]
+    pub fn selected_token(&self) -> Option<LlamaToken> {
+        self.data
+            .get(usize::try_from(self.selected).ok()?)
+            .map(LlamaTokenData::id)
+    }
+}
+
+impl LlamaTokenDataArray {
+    /// Modify the underlying data as a `llama_token_data_array`. and reconstruct the `LlamaTokenDataArray`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if some of the safety conditions are not met. (we cannot check all of them at runtime so breaking them is UB)
+    ///
+    /// SAFETY:
+    /// [modify] cannot change the data pointer.
+    /// if the data is not sorted, sorted must be false.
+    /// the size of the data can only decrease (i.e you cannot add new elements).
+    pub(crate) unsafe fn modify_as_c_llama_token_data_array<T>(
+        &mut self,
+        modify: impl FnOnce(&mut llama_cpp_sys_2::llama_token_data_array) -> T,
+    ) -> T {
+        let size = self.data.len();
+        let data = self.data.as_mut_ptr().cast();
+        let mut c_llama_token_data_array = llama_cpp_sys_2::llama_token_data_array {
+            data,
+            size,
+            selected: self.selected,
+            sorted: self.sorted,
+        };
+        let result = modify(&mut c_llama_token_data_array);
+        assert!(
+            ptr::eq(data, c_llama_token_data_array.data),
+            "data pointer changed"
+        );
+        assert!(c_llama_token_data_array.size <= size, "size increased");
+        self.data.set_len(c_llama_token_data_array.size);
+        self.sorted = c_llama_token_data_array.sorted;
+        self.selected = c_llama_token_data_array.selected;
+        result
+    }
+
+    pub(crate) unsafe fn apply_sampler(&mut self, sampler: *mut llama_cpp_sys_2::llama_sampler) {
+        self.modify_as_c_llama_token_data_array(|c_llama_token_data_array| {
+            llama_cpp_sys_2::llama_sampler_apply(sampler, c_llama_token_data_array);
+        })
+    }
+
+    pub(crate) unsafe fn apply_and_free_sampler(
+        &mut self,
+        sampler_fn: impl FnOnce() -> *mut llama_cpp_sys_2::llama_sampler,
+    ) {
+        let sampler = sampler_fn();
+        self.apply_sampler(sampler);
+        llama_cpp_sys_2::llama_sampler_free(sampler);
+    }
```

Comment:
I know this is pub(crate) but I think a repr(transparent) struct with `Drop` would make maintaining this abstraction easier.